File: packages/core/src/openseed_core/harness/generator.py
"""Harness scaffold generator — deterministic template generation from scan results."""

from __future__ import annotations

import json
import os
from dataclasses import dataclass, field
from pathlib import Path


@dataclass
class ScanResult:
    """Result of project folder scan."""

    root: str
    name: str
    languages: list[str] = field(default_factory=list)
    frameworks: list[str] = field(default_factory=list)
    package_manager: str | None = None
    test_runner: str | None = None
    linter: str | None = None
    linter_config: str | None = None
    formatter: str | None = None
    type_checker: str | None = None
    build_tool: str | None = None
    is_monorepo: bool = False
    monorepo_tool: str | None = None
    packages: list[PackageInfo] = field(default_factory=list)
    commands: dict[str, str] = field(default_factory=dict)


@dataclass
class PackageInfo:
    name: str
    path: str
    description: str


@dataclass
class HarnessFile:
    """A file to be written as part of harness scaffold."""

    path: str  # relative to project root
    content: str


def scan_project(working_dir: str) -> ScanResult:
    """Scan a project directory and detect tech stack. Deterministic, no AI."""
    root = Path(working_dir)
    name = root.name
    result = ScanResult(root=working_dir, name=name)

    # Python
    pyproject = root / "pyproject.toml"
    if pyproject.is_file():
        content = pyproject.read_text(errors="ignore")
        result.languages.append("Python")
        if "uv" in content.lower():
            result.package_manager = "uv"
        elif (root / "Pipfile").exists():
            result.package_manager = "pipenv"
        elif (root / "poetry.lock").exists():
            result.package_manager = "poetry"
        else:
            result.package_manager = "pip"

        if "ruff" in content:
            result.linter = "ruff"
            result.linter_config = "pyproject.toml"
            result.formatter = "ruff"
        if "mypy" in content:
            result.type_checker = "mypy"
        if "pytest" in content:
            result.test_runner = "pytest"
        if "workspace" in content.lower():
            result.is_monorepo = True
            result.monorepo_tool = "uv workspace"

    elif (root / "requirements.txt").exists():
        result.languages.append("Python")
        result.package_manager = "pip"

    # Node.js / TypeScript
    pkg_json = root / "package.json"
    if pkg_json.is_file():
        try:
            data = json.loads(pkg_json.read_text())
            deps = {**data.get("dependencies", {}), **data.get("devDependencies", {})}

            if "typescript" in deps or (root / "tsconfig.json").exists():
                result.languages.append("TypeScript")
            else:
                result.languages.append("JavaScript")

            # Package manager
            if (root / "pnpm-lock.yaml").exists():
                result.package_manager = result.package_manager or "pnpm"
            elif (root / "yarn.lock").exists():
                result.package_manager = result.package_manager or "yarn"
            elif (root / "bun.lockb").exists():
                result.package_manager = result.package_manager or "bun"
            else:
                result.package_manager = result.package_manager or "npm"

            # Frameworks
            for name_check, label in [
                ("next", "Next.js"), ("react", "React"), ("vue", "Vue"),
                ("express", "Express"), ("fastify", "Fastify"), ("hono", "Hono"),
                ("astro", "Astro"), ("svelte", "Svelte"),
            ]:
                if name_check in deps:
                    result.frameworks.append(label)

            # Test runner
            if not result.test_runner:
                if "vitest" in deps:
                    result.test_runner = "vitest"
                elif "jest" in deps:
                    result.test_runner = "jest"

            # Linter
            if not result.linter:
                if "@biomejs/biome" in deps:
                    result.linter = "biome"
                    result.linter_config = "biome.json"
                elif "eslint" in deps:
                    result.linter = "eslint"

            # Monorepo
            if "workspaces" in data:
                result.is_monorepo = True
            if (root / "turbo.json").exists():
                result.is_monorepo = True
                result.monorepo_tool = "turborepo"
            elif (root / "nx.json").exists():
                result.is_monorepo = True
                result.monorepo_tool = "nx"
            elif (root / "pnpm-workspace.yaml").exists():
                result.is_monorepo = True
                result.monorepo_tool = "pnpm workspaces"

            # Scripts → commands
            scripts = data.get("scripts", {})
            pm = result.package_manager or "npm"
            run_prefix = "npm run" if pm == "npm" else pm
            for script_name, cmd_key in [
                ("build", "build"), ("dev", "dev"), ("test", "test"),
                ("lint", "lint"), ("typecheck", "typecheck"),
            ]:
                if script_name in scripts:
                    result.commands[cmd_key] = f"{run_prefix} {script_name}"
        except Exception:
            result.languages.append("JavaScript")

    # Go
    if (root / "go.mod").exists():
        result.languages.append("Go")
        result.package_manager = result.package_manager or "go mod"
        result.test_runner = result.test_runner or "go test"

    # Rust
    if (root / "Cargo.toml").exists():
        result.languages.append("Rust")
        result.package_manager = result.package_manager or "cargo"
        result.test_runner = result.test_runner or "cargo test"

    # Build commands from detected tools
    if not result.commands:
        result.commands = _build_commands(result)

    # Detect monorepo packages
    if result.is_monorepo:
        result.packages = _detect_packages(root)

    return result


def _build_commands(scan: ScanResult) -> dict[str, str]:
    """Generate key commands from detected tools."""
    cmds: dict[str, str] = {}
    pm = scan.package_manager

    if pm == "uv":
        cmds["install"] = "uv sync"
    elif pm in ("pnpm", "yarn", "npm", "bun"):
        cmds["install"] = f"{pm} install"
    elif pm == "pip":
        cmds["install"] = "pip install -r requirements.txt"
    elif pm == "poetry":
        cmds["install"] = "poetry install"

    if scan.test_runner:
        cmds["test"] = scan.test_runner
    if scan.linter:
        if scan.linter == "ruff":
            cmds["lint"] = "ruff check ."
        elif scan.linter == "biome":
            cmds["lint"] = "biome check ."
        elif scan.linter == "eslint":
            cmds["lint"] = "eslint ."
    if scan.formatter:
        if scan.formatter == "ruff":
            cmds["format"] = "ruff format ."
    if scan.type_checker:
        cmds["typecheck"] = f"{scan.type_checker} ."

    return cmds


def _detect_packages(root: Path) -> list[PackageInfo]:
    """Detect monorepo sub-packages."""
    packages: list[PackageInfo] = []
    for container in ["packages", "apps", "services", "libs", "modules"]:
        container_path = root / container
        if not container_path.is_dir():
            continue
        for entry in sorted(container_path.iterdir()):
            if entry.is_dir() and not entry.name.startswith("."):
                desc = _infer_description(entry.name)
                packages.append(PackageInfo(
                    name=entry.name,
                    path=f"{container}/{entry.name}",
                    description=desc,
                ))
    return packages


def _infer_description(name: str) -> str:
    hints = {
        "web": "web frontend", "app": "application", "api": "API server",
        "core": "core business logic", "shared": "shared utilities",
        "types": "shared types", "ui": "UI components", "db": "database layer",
        "auth": "authentication", "cli": "CLI tool", "server": "backend server",
    }
    return hints.get(name, f"{name} package")


def generate_scaffold(scan: ScanResult, existing_files: set[str] | None = None) -> list[HarnessFile]:
    """Generate harness scaffold files from scan result. Deterministic, no AI.

    Only generates files that don't already exist (checks existing_files set
    or defaults to checking the filesystem).
    """
    import os

    if existing_files is None:
        existing_files = set()
        for candidate in ["AGENTS.md", ".pre-commit-config.yaml", ".github/workflows/ci.yml"]:
            if os.path.exists(os.path.join(scan.root, candidate)):
                existing_files.add(candidate)

    files: list[HarnessFile] = []

    # ── Inform ──────────────────────────────────────────────────

    # Root AGENTS.md
    if "AGENTS.md" not in existing_files:
        files.append(HarnessFile(
            path="AGENTS.md",
            content=_generate_root_agents_md(scan),
        ))

    # Sub-AGENTS.md for monorepo packages
    if scan.is_monorepo:
        for pkg in scan.packages:
            pkg_agents = f"{pkg.path}/AGENTS.md"
            if pkg_agents not in existing_files:
                files.append(HarnessFile(
                    path=pkg_agents,
                    content=_generate_sub_agents_md(pkg, scan),
                ))

    # ── Constrain ───────────────────────────────────────────────

    # Pre-commit config
    if ".pre-commit-config.yaml" not in existing_files:
        precommit = _generate_precommit_config(scan)
        if precommit:
            files.append(HarnessFile(
                path=".pre-commit-config.yaml",
                content=precommit,
            ))

    # ── Verify ──────────────────────────────────────────────────

    # CI pipeline
    if ".github/workflows/ci.yml" not in existing_files:
        ci = _generate_ci_pipeline(scan)
        if ci:
            files.append(HarnessFile(
                path=".github/workflows/ci.yml",
                content=ci,
            ))

    return files


def _generate_root_agents_md(scan: ScanResult) -> str:
    sections: list[str] = []

    # Mission (placeholder for AI to fill)
    sections.append("# AGENTS.md\n")
    sections.append(f"> **Project:** {scan.name} — [TODO: describe project in 1-2 sentences]")
    sections.append("")

    # Key Commands
    sections.append("## Key Commands")
    sections.append("| Intent | Command | Notes |")
    sections.append("|--------|---------|-------|")
    for intent, cmd in scan.commands.items():
        note = ""
        if intent == "lint" and scan.linter_config:
            note = f"see {scan.linter_config}"
        elif intent == "typecheck" and scan.type_checker:
            note = f"{scan.type_checker}"
        sections.append(f"| {intent.capitalize()} | `{cmd}` | {note} |")
    sections.append("")

    # Architecture Constraints
    if scan.is_monorepo and scan.packages:
        sections.append("## Architecture Constraints")
        pkg_names = [p.name for p in scan.packages]
        sections.append(f"- Packages: {', '.join(pkg_names)}")
        sections.append("- [TODO: define dependency flow between packages]")
        sections.append("")

    # Code Style
    sections.append("## Code Style")
    for lang in scan.languages:
        if lang == "Python":
            sections.append("- Type hints on all public functions")
            sections.append("- async/await for I/O operations")
        elif lang in ("TypeScript", "JavaScript"):
            sections.append("- Named exports preferred")
        elif lang == "Go":
            sections.append("- Always handle errors explicitly")
    sections.append("- Structured logging only. No print/console.log in production.")
    sections.append("")

    # Boundaries
    sections.append("## Boundaries\n")
    sections.append("### NEVER")
    sections.append("- Commit secrets, tokens, or .env files")
    sections.append("- Force push to main")
    sections.append("")
    sections.append("### ASK")
    sections.append("- Before adding new external dependencies")
    sections.append("")
    sections.append("### ALWAYS")
    verify_parts = []
    if scan.commands.get("lint"):
        verify_parts.append(scan.commands["lint"])
    if scan.commands.get("typecheck"):
        verify_parts.append(scan.commands["typecheck"])
    if scan.commands.get("test"):
        verify_parts.append(scan.commands["test"])
    if verify_parts:
        sections.append(f"- Run `{' && '.join(verify_parts)}` before marking task complete")
    sections.append("- Handle all errors explicitly")
    sections.append("")

    # Context Map
    if scan.is_monorepo and scan.packages:
        sections.append("## Context Map")
        sections.append("```yaml")
        if scan.monorepo_tool:
            sections.append(f"monorepo: {scan.monorepo_tool}")
        sections.append("\npackages:")
        for pkg in scan.packages:
            sections.append(f"  {pkg.path}: {pkg.description}")
        sections.append("```")
        sections.append("")

    return "\n".join(sections)


def _generate_sub_agents_md(pkg: PackageInfo, scan: ScanResult) -> str:
    sections: list[str] = []
    sections.append(f"# AGENTS.md ({pkg.path}/)\n")
    sections.append("## Scope")
    sections.append(f"{pkg.description}\n")
    sections.append("## Rules")
    sections.append("- [TODO: add package-specific rules]\n")
    sections.append("## Testing")

    test_cmd = scan.commands.get("test", "pytest")
    if "Python" in scan.languages:
        sections.append(f"- Run: `{test_cmd} {pkg.path}/tests/`")
    else:
        pm = scan.package_manager or "npm"
        sections.append(f"- Run: `{pm} --filter {pkg.name} test`")

    return "\n".join(sections)


def _generate_precommit_config(scan: ScanResult) -> str | None:
    """Generate .pre-commit-config.yaml based on detected toolchain."""
    hooks: list[str] = []

    if "Python" in scan.languages:
        if scan.linter == "ruff":
            hooks.append("""  - repo: https://github.com/astral-sh/ruff-pre-commit
    rev: v0.11.6
    hooks:
      - id: ruff
        args: [check, --fix]
      - id: ruff-format""")
        if scan.type_checker == "mypy":
            hooks.append("""  - repo: local
    hooks:
      - id: mypy
        name: mypy
        entry: mypy .
        language: system
        types: [python]
        pass_filenames: false""")

    if "TypeScript" in scan.languages or "JavaScript" in scan.languages:
        if scan.linter == "biome":
            hooks.append("""  - repo: local
    hooks:
      - id: biome
        name: biome check
        entry: biome check --write
        language: system
        types_or: [javascript, jsx, ts, tsx, json]""")
        elif scan.linter == "eslint":
            hooks.append("""  - repo: local
    hooks:
      - id: eslint
        name: eslint
        entry: eslint --fix
        language: system
        types_or: [javascript, jsx, ts, tsx]""")

    if not hooks:
        return None

    return "repos:\n" + "\n".join(hooks) + "\n"


def _generate_ci_pipeline(scan: ScanResult) -> str | None:
    """Generate .github/workflows/ci.yml based on detected toolchain."""
    steps: list[str] = []

    steps.append("      - uses: actions/checkout@v4")

    if "Python" in scan.languages:
        pm = scan.package_manager or "pip"
        if pm == "uv":
            steps.append("""
      - uses: astral-sh/setup-uv@v4
        with:
          version: "latest"

      - uses: actions/setup-python@v5
        with:
          python-version: "3.11"

      - name: Install dependencies
        run: uv sync""")
        else:
            steps.append("""
      - uses: actions/setup-python@v5
        with:
          python-version: "3.11"

      - name: Install dependencies
        run: pip install -r requirements.txt""")

        run_prefix = "uv run " if pm == "uv" else ""

        if scan.commands.get("lint"):
            steps.append(f"""
      - name: Lint
        run: {run_prefix}{scan.commands['lint']}""")
        if scan.commands.get("format"):
            steps.append(f"""
      - name: Format check
        run: {run_prefix}{scan.commands['format']} --check""")
        if scan.commands.get("typecheck"):
            steps.append(f"""
      - name: Type check
        run: {run_prefix}{scan.commands['typecheck']}""")
        if scan.commands.get("test"):
            steps.append(f"""
      - name: Test
        run: {run_prefix}{scan.commands['test']}""")

    elif "TypeScript" in scan.languages or "JavaScript" in scan.languages:
        pm = scan.package_manager or "npm"
        steps.append(f"""
      - uses: actions/setup-node@v4
        with:
          node-version: "20"

      - name: Install dependencies
        run: {pm} install""")

        for intent in ["lint", "typecheck", "test"]:
            if scan.commands.get(intent):
                steps.append(f"""
      - name: {intent.capitalize()}
        run: {scan.commands[intent]}""")

    elif "Go" in scan.languages:
        steps.append("""
      - uses: actions/setup-go@v5
        with:
          go-version: "stable"

      - name: Test
        run: go test ./...""")

    if len(steps) <= 1:  # Only checkout, no meaningful steps
        return None

    return f"""name: CI

on:
  pull_request:
    branches: [main]
  push:
    branches: [main]

jobs:
  lint-and-test:
    runs-on: ubuntu-latest
    steps:
{"".join(steps)}
"""


def get_ai_guide() -> str:
    """Load the AI harness guide for use as LLM context."""
    guide_path = Path(__file__).parent / "ai_guide.md"
    return guide_path.read_text(encoding="utf-8")

File: harness-bootstrap/src/generator.ts
import type {
  ProjectAnalysis,
  HarnessOutput,
  DocsFile,
  SubAgentsMd,
  CurationItem,
} from "./types.js";

// ── Harness Generator (Phase 1 Output) ─────────────────────

export function generateHarness(
  analysis: ProjectAnalysis,
  curationAnswers: Map<string, string>
): HarnessOutput {
  const agentsMd = generateAgentsMd(analysis, curationAnswers);
  const globalAgentsMd = generateGlobalAgentsMd(analysis);
  const configToml = generateConfigToml();
  const docsStructure = generateDocsStructure(analysis, curationAnswers);
  const subAgentsMd = generateSubAgentsMd(analysis);

  return {
    agentsMd,
    globalAgentsMd,
    configToml,
    docsStructure,
    subAgentsMd,
    claudeMdSymlink: true,
  };
}

// ── AGENTS.md Generation ────────────────────────────────────

function generateAgentsMd(
  analysis: ProjectAnalysis,
  answers: Map<string, string>
): string {
  const sections: string[] = [];
  const { techStack: stack, commands, monorepo } = analysis;

  // ── Mission
  const mission = answers.get("mission") ?? `[TODO: Describe your project in 1-2 sentences]`;
  const coreConstraint = answers.get("core-constraint");

  sections.push(`# AGENTS.md\n`);
  sections.push(`> **Project:** ${mission}`);
  if (coreConstraint && coreConstraint !== "없음") {
    sections.push(`> **Core constraint:** ${coreConstraint}`);
  }
  sections.push("");

  // ── Key Commands
  sections.push(`## Key Commands`);
  sections.push(`| Intent | Command | Notes |`);
  sections.push(`|--------|---------|-------|`);

  const commandEntries: [string, string | null, string][] = [
    ["Install", commands.install, stack.packageManager ? `${stack.packageManager} only` : ""],
    ["Build", commands.build, stack.buildTool ? `${stack.buildTool}` : ""],
    ["Dev", commands.dev, ""],
    ["Test", commands.test, stack.testRunner ?? ""],
    ["Lint", commands.lint, stack.linter ? `${stack.linter.name} (see \`${stack.linter.configFile}\`)` : ""],
    ["Type check", commands.typecheck, stack.languages.includes("TypeScript") ? "tsc --noEmit" : ""],
    ["Format", commands.format, stack.formatter ?? ""],
    ["E2E", commands.e2e, ""],
    ["DB migrate", commands.migrate, stack.orm ?? ""],
  ];

  for (const [intent, cmd, notes] of commandEntries) {
    if (cmd) {
      sections.push(`| ${intent} | \`${cmd}\` | ${notes} |`);
    }
  }
  sections.push("");

  // ── Architecture Constraints
  if (monorepo && monorepo.packages.length > 1) {
    sections.push(`## Architecture Constraints`);
    const depFlow = answers.get("dependency-flow");
    if (depFlow) {
      sections.push(`- Dependency flow: ${depFlow}`);
    }
    const nonObvious = answers.get("non-obvious-patterns");
    if (nonObvious && nonObvious !== "없음") {
      sections.push(`- ${nonObvious}`);
    }
    sections.push("");
  }

  // ── Code Style
  sections.push(`## Code Style`);
  if (stack.languages.includes("TypeScript")) {
    sections.push(`- TypeScript strict mode. No \`any\`, no \`as\` assertions without justifying comment.`);
    sections.push(`- Named exports only. No default exports except framework page components.`);
  }

  const errorHandling = answers.get("error-handling");
  if (errorHandling && errorHandling !== "프로젝트에 맞게 AI가 추천") {
    sections.push(`- Error handling: ${errorHandling}`);
  }
  sections.push(`- Logging: structured JSON. No console.log in production code.`);
  sections.push("");

  // ── Non-Obvious Patterns
  const nonObvious = answers.get("non-obvious-patterns");
  if (nonObvious && nonObvious !== "없음") {
    sections.push(`## Non-Obvious Patterns`);
    sections.push(`- ${nonObvious}`);
    sections.push("");
  }

  // ── Testing Rules
  if (stack.testRunner) {
    sections.push(`## Testing Rules`);
    const coverage = answers.get("test-coverage");
    if (coverage && coverage !== "커버리지 요구 없음") {
      sections.push(`- Minimum coverage: ${coverage}`);
    }
    sections.push(`- All tests must be deterministic and isolated. Mock external dependencies.`);
    if (commands.test) {
      sections.push(`- Run \`${commands.test}\` before marking any task complete.`);
    }
    sections.push("");
  }

  // ── Boundaries
  sections.push(`## Boundaries\n`);

  // NEVER
  sections.push(`### NEVER`);
  sections.push(`- Commit secrets, tokens, or .env files`);
  sections.push(`- Force push to main or protected branches`);
  sections.push(`- Modify vendor/, dist/, or build/ directories`);
  const neverExtra = answers.get("never-boundaries");
  if (neverExtra && neverExtra !== "기본값만 사용") {
    sections.push(`- ${neverExtra}`);
  }
  sections.push("");

  // ASK
  sections.push(`### ASK`);
  sections.push(`- Before adding new external dependencies`);
  sections.push(`- Before deleting files`);
  const askExtra = answers.get("ask-boundaries");
  if (askExtra && askExtra !== "기본값만 사용") {
    sections.push(`- ${askExtra}`);
  }
  sections.push("");

  // ALWAYS
  sections.push(`### ALWAYS`);
  sections.push(`- Explain your plan before writing code`);
  sections.push(`- Handle all errors explicitly — never silently suppress exceptions`);

  const verifyCommands: string[] = [];
  if (commands.lint) verifyCommands.push(commands.lint);
  if (commands.typecheck) verifyCommands.push(commands.typecheck);
  if (commands.test) verifyCommands.push(commands.test);
  if (verifyCommands.length > 0) {
    sections.push(`- Run \`${verifyCommands.join(" && ")}\` before marking task complete`);
  }
  sections.push("");

  // ── Persona
  const persona = answers.get("persona");
  if (persona && persona !== "페르소나 없음") {
    if (persona.includes("멀티")) {
      sections.push(`## Personas`);
      sections.push(`Invoke via skill: @Lead, @Dev, @Critic`);
      sections.push(`Definitions: \`.claude/skills/\``);
    } else {
      sections.push(`## Identity`);
      sections.push(`${persona} — ${stack.languages.join(", ")}${stack.frameworks.length ? `, ${stack.frameworks.join(", ")}` : ""}.`);
      sections.push(`Favor explicit error handling and composition over inheritance.`);
    }
    sections.push("");
  }

  // ── Git Conventions
  const gitConv = answers.get("git-conventions");
  if (gitConv && gitConv !== "Free-form") {
    sections.push(`## Git Conventions`);
    if (gitConv.includes("Conventional")) {
      sections.push(`- Commit format: \`type(scope): description\``);
      sections.push(`  Types: feat, fix, refactor, test, docs, chore, perf`);
    } else if (gitConv.includes("Gitmoji")) {
      sections.push(`- Commit format: \`:emoji: description\` (Gitmoji convention)`);
    }
    sections.push(`- PR must pass CI before merge. Squash merge only.`);
    sections.push(`- Branch naming: \`type/short-description\``);
    sections.push("");
  }

  // ── Context Map
  if (monorepo || Object.keys(analysis.structure.notable).length > 3) {
    sections.push(`## Context Map`);
    sections.push("```yaml");
    if (monorepo) {
      sections.push(`monorepo: ${monorepo.tool}\n`);
      sections.push(`packages:`);
      for (const pkg of monorepo.packages) {
        sections.push(`  ${pkg.path}: ${pkg.description}`);
      }
    }
    if (Object.keys(analysis.structure.notable).length > 0) {
      sections.push(`\nnotable:`);
      for (const [dir, desc] of Object.entries(analysis.structure.notable)) {
        if (!["node_modules", "dist", "build", ".git"].includes(dir)) {
          sections.push(`  ${dir}/: ${desc}`);
        }
      }
    }
    sections.push("```");
    sections.push("");
  }

  return sections.join("\n");
}

// ── Global AGENTS.md ────────────────────────────────────────

function generateGlobalAgentsMd(analysis: ProjectAnalysis): string {
  const { techStack: stack } = analysis;
  const sections: string[] = [];

  sections.push(`# Global AGENTS.md\n`);
  sections.push(`## General Preferences`);

  if (stack.languages.includes("TypeScript")) {
    sections.push(`- Language: TypeScript (prefer over JavaScript in all cases)`);
  }
  if (stack.packageManager) {
    sections.push(`- Package manager: ${stack.packageManager} (always use ${stack.packageManager})`);
  }
  sections.push(`- Always run lint and typecheck before considering a task done`);
  sections.push(`- Commit messages in English, conventional commit format`);
  sections.push("");

  sections.push(`## Security (Global)`);
  sections.push(`- Never write secrets, API keys, or credentials into any file`);
  sections.push(`- Never disable linter rules with inline comments without justification`);
  sections.push("");

  sections.push(`## Quality`);
  sections.push(`- No TODO comments without a linked issue number`);
  sections.push(`- Remove unused imports and dead code before committing`);
  sections.push(`- Prefer composition over inheritance`);
  sections.push(`- Keep functions under 40 lines when possible`);

  return sections.join("\n");
}

// ── config.toml ─────────────────────────────────────────────

function generateConfigToml(): string {
  return `# ~/.codex/config.toml
project_doc_fallback_filenames = ["TEAM_GUIDE.md", ".agents.md"]
project_doc_max_bytes = 65536
`;
}

// ── docs/ Structure ─────────────────────────────────────────

function generateDocsStructure(
  analysis: ProjectAnalysis,
  answers: Map<string, string>
): DocsFile[] {
  const files: DocsFile[] = [];

  // architecture/overview.md
  const mission = answers.get("mission") ?? "[TODO]";
  const constraint = answers.get("core-constraint");

  files.push({
    path: "docs/architecture/overview.md",
    content: `# Architecture Overview

## Project
${mission}
${constraint && constraint !== "없음" ? `\n## Core Constraint\n${constraint}` : ""}

## Tech Stack
${analysis.techStack.languages.map((l) => `- ${l}`).join("\n")}
${analysis.techStack.frameworks.map((f) => `- ${f}`).join("\n")}
${analysis.techStack.database ? `- Database: ${analysis.techStack.database}` : ""}
${analysis.techStack.orm ? `- ORM: ${analysis.techStack.orm}` : ""}

## Module Structure
${
  analysis.monorepo
    ? analysis.monorepo.packages.map((p) => `- \`${p.path}\`: ${p.description}`).join("\n")
    : Object.entries(analysis.structure.notable)
        .map(([dir, desc]) => `- \`${dir}/\`: ${desc}`)
        .join("\n")
}

## Dependency Rules
${answers.get("dependency-flow") ?? "[TODO: Define dependency flow between modules]"}
`,
  });

  // architecture/dependency-graph.md
  files.push({
    path: "docs/architecture/dependency-graph.md",
    content: `# Dependency Graph

## Rules
${answers.get("dependency-flow") ?? "[TODO: Define allowed import directions]"}

## Enforcement
- Structural tests validate compliance on every PR
- CI blocks merges that violate dependency rules
- Linter rules prevent cross-boundary imports

## Visualization
[TODO: Add mermaid diagram or ASCII art of dependency flow]
`,
  });

  // architecture/adr/000-template.md
  files.push({
    path: "docs/architecture/adr/000-template.md",
    content: `# ADR-000: [Title]

## Status
[Proposed | Accepted | Deprecated | Superseded]

## Context
[What is the issue that we're seeing that is motivating this decision?]

## Decision
[What is the change that we're proposing and/or doing?]

## Consequences
[What becomes easier or more difficult to do because of this change?]

## Alternatives Considered
[What other options did we evaluate?]
`,
  });

  // maps/module-map.md
  files.push({
    path: "docs/maps/module-map.md",
    content: `# Module Map

## Overview
This map helps agents navigate the codebase efficiently.

## Key Directories
${Object.entries(analysis.structure.notable)
  .map(([dir, desc]) => `### \`${dir}/\`\n${desc}\n`)
  .join("\n")}

## Entry Points
[TODO: List main entry points for the application]

## Key Files
[TODO: List critical files that agents should understand]
`,
  });

  // maps/execution-plan.md
  files.push({
    path: "docs/maps/execution-plan.md",
    content: `# Agent Execution Plan

## Workflow
1. Read AGENTS.md and relevant docs/
2. Analyze the task requirements
3. Check existing tests and patterns
4. Implement changes following architectural constraints
5. Run verification: lint → typecheck → test
6. Self-review changes before submitting

## Verification Checklist
${analysis.commands.lint ? `- [ ] Lint passes: \`${analysis.commands.lint}\`` : ""}
${analysis.commands.typecheck ? `- [ ] Type check passes: \`${analysis.commands.typecheck}\`` : ""}
${analysis.commands.test ? `- [ ] Tests pass: \`${analysis.commands.test}\`` : ""}
- [ ] No new linter warnings
- [ ] No unused imports or dead code
- [ ] Error handling is explicit

## Escalation
- If blocked for more than 3 attempts, pause and report the issue
- If architectural decision is needed, create an ADR draft in docs/adr/
`,
  });

  // conventions/ (language-specific)
  if (analysis.techStack.languages.includes("TypeScript")) {
    files.push({
      path: "docs/conventions/typescript-style.md",
      content: `# TypeScript Conventions

## General
- Strict mode enabled (see tsconfig.json)
- No \`any\` type. Use \`unknown\` and narrow.
- No \`as\` type assertions without justifying comment.
- Named exports only. No default exports except framework pages.

## Naming
- camelCase for variables, functions, methods
- PascalCase for types, interfaces, classes, components
- UPPER_SNAKE_CASE for constants
- Prefix interfaces with context, not \`I\` (e.g., \`UserRepository\` not \`IUserRepository\`)

## Error Handling
${answers.get("error-handling") ?? "- Use domain-specific error classes"}
- Never swallow errors silently
- Log all unhandled errors with structured logging

## Async
- Prefer async/await over .then() chains
- Always handle promise rejections
- Use AbortController for cancellable operations

## Imports
- Group: external → internal → relative
- No circular imports (enforced by linter)
`,
    });
  }

  // conventions/testing-strategy.md
  if (analysis.techStack.testRunner) {
    files.push({
      path: "docs/conventions/testing-strategy.md",
      content: `# Testing Strategy

## Test Runner
${analysis.techStack.testRunner}

## Principles
- Tests must be deterministic and isolated
- Mock external dependencies (network, DB, filesystem)
- Test behavior, not implementation
- Each test should have a single assertion focus

## Coverage
${answers.get("test-coverage") ?? "[TODO: Define coverage requirements]"}

## Structure
- Unit tests: colocated with source or in __tests__/
- Integration tests: tests/integration/
- E2E tests: ${analysis.commands.e2e ? `\`${analysis.commands.e2e}\`` : "tests/e2e/"}

## Naming
- Describe behavior: \`it("returns 404 when user not found")\`
- Group by module/function: \`describe("UserService")\`
`,
    });
  }

  return files;
}

// ── Sub-directory AGENTS.md ─────────────────────────────────

function generateSubAgentsMd(analysis: ProjectAnalysis): SubAgentsMd[] {
  const subs: SubAgentsMd[] = [];

  if (!analysis.monorepo) return subs;

  for (const pkg of analysis.monorepo.packages) {
    const isApi = /api|server|backend/.test(pkg.name);
    const isWeb = /web|app|frontend|client/.test(pkg.name);
    const isDb = /db|database|prisma/.test(pkg.name);
    const isCore = /core|shared|common|lib/.test(pkg.name);
    const isTypes = /types|schema/.test(pkg.name);

    let content = `# AGENTS.md (${pkg.path}/)\n\n## Scope\n${pkg.description}\n\n## Rules\n`;

    if (isApi) {
      content += `- Every route file exports a framework plugin/handler
- Request/response schemas validated at boundary
- No business logic here; delegate to core/services
- All responses follow consistent envelope format
- Test both success and error paths for every endpoint

## Testing
- Run: \`${analysis.techStack.packageManager ?? "npm"} --filter ${pkg.name} test\`
- Use integration tests with mock external services
`;
    } else if (isWeb) {
      content += `- Components must be pure/functional where possible
- No direct API calls from components; use hooks/services
- Colocate component tests with component files
- Use semantic HTML elements
- Accessibility: all interactive elements must be keyboard navigable

## Testing
- React Testing Library for component tests. No snapshot tests.
- Test user interactions, not implementation details
`;
    } else if (isDb) {
      content += `- All database queries live here as repository functions
- Use parameterized queries only; no string concatenation
- New tables require an ADR in docs/adr/ before creation
- Never edit existing migration files; create new ones

## Migrations
- Migration names: YYYYMMDD_description
- Always test migrations up and down
`;
    } else if (isCore) {
      content += `- Pure business logic only; no framework-specific imports
- No side effects in pure functions
- All public functions must have JSDoc documentation
- Maximum test coverage priority

## Testing
- Unit tests for all public API methods
- Aim for highest coverage in the monorepo
`;
    } else if (isTypes) {
      content += `- Types must be framework-agnostic
- Use branded types for domain identifiers (UserId, OrderId, etc.)
- Export all types via barrel file (index.ts)
- No runtime code in this package; types only
`;
    } else {
      content += `- [TODO: Add package-specific rules]\n`;
    }

    subs.push({ path: `${pkg.path}/AGENTS.md`, content });
  }

  return subs;
}

// ── Default Answers (for fully automated mode) ──────────────

export function getDefaultAnswers(): Map<string, string> {
  return new Map([
    ["mission", "[TODO: Describe your project]"],
    ["core-constraint", "없음"],
    ["non-obvious-patterns", "없음"],
    ["never-boundaries", "기본값만 사용"],
    ["ask-boundaries", "기본값만 사용"],
    ["error-handling", "프로젝트에 맞게 AI가 추천"],
    ["test-coverage", "80% line coverage"],
    ["persona", "페르소나 없음"],
    ["git-conventions", "Conventional Commits (feat/fix/refactor/test/docs/chore)"],
  ]);
}
